### PR TITLE
Laundry refactor

### PR DESCRIPTION
**Pull request by adelq**
Much easier to read and has more rigorous testing now.

### CODE REVIEW

File: penn/laundry.py [hunk 1]
```diff
@@ -25,37 +29,41 @@ def all_status(self):
...
+            row_data = (val.get_text().strip() for val in row.find_all('td'))
+            clean_row = [val for val in row_data if len(val) > 0]
+            data.append(clean_row)
+
+        # Remove the header row and all empty rows
+        data_improved = [row for row in data if len(row) > 0][1:]
+
+        # Construct the final JSON
+        laundry_rooms = []
+        for row in data_improved[1:]:
```
Comment (dlakata) [inline penn/laundry.py]: any reason why ```data_improved``` is sliced both here and on 53?

Comment (adelq) [inline penn/laundry.py]: Not sure actually, this was inherited from the original codebase. Will give you a more concrete answer momentarily.

Comment (adelq) [inline penn/laundry.py]: So it appears that these rows correspond to the following:
```python
[u'Harrison 12th FL not reporting status. \xa0Service has been requested.']
[u'LAUNDRY ROOM', u'WASHERSAVAILABLE', u'DRYERSAVAILABLE', u'WASHERSIN USE', u'DRYERSIN USE']
```
It looks like one of them was an intentional header removal, the other was to fix the addition of a row to the markup for a broken washer, which has not been resolved. Once this row is removed however, we should only be slicing the first row. I've added another commit to clarify this change.

File: penn/laundry.py [hunk 2]
```diff
@@ -25,37 +29,41 @@ def all_status(self):
...
-                room_dict = dict()
-                room_dict['washers_available'] = int(row[1])
-                room_dict['dryers_available'] = int(row[2])
-                room_dict['washers_in_use'] = int(row[3])
-                room_dict['dryers_in_use'] = int(row[4])
-                room_dict['hall_no'] = hall_dict[row[0]]
-                room_dict['name'] = row[0]
-                laundry_rooms.append(room_dict)
-            return laundry_rooms
+        if not r.status_code == 200:
```
Comment (dlakata) [inline penn/laundry.py]: ```if r.status_code != 200```

File: penn/laundry.py [hunk 3]
```diff
@@ -73,25 +81,33 @@ def hall_status(self, hall_no):
...
-                d['available'] = data_row[2] == u'Available'
-                if len(data_row) == 4:
-                    d['time_left'] = data_row[3]
-                else:
-                    d['time_left'] = None
-                return d
-
-            return {'machines': list(map(toDict, data_improved)), 'hall_name': hall_name}
-        else:
+        if not r.status_code == 200:
```
Comment (dlakata) [inline penn/laundry.py]: ```if r.status_code != 200```

File: penn/laundry.py (hunk 2 above)
Comment (adelq) [inline penn/laundry.py]: Now that I think about it, this should not be returning a dictionary, it should raise an exception and be caught by the mobile server. Will use `r.raise_for_status()`.

**Comment by afischer** [comment]
2013-08-23-tumblr_m08celcrnt1r3isxuo1_500
On it.

File: penn/laundry.py [hunk 4]
```diff
@@ -73,25 +80,32 @@ def hall_status(self, hall_no):
...
-                if len(data_row) == 4:
-                    d['time_left'] = data_row[3]
-                else:
-                    d['time_left'] = None
-                return d
-
-            return {'machines': list(map(toDict, data_improved)), 'hall_name': hall_name}
-        else:
-            return {'error': 'The laundry api is currently unavailable.'}
+        r.raise_for_status()
```
Comment (afischer) [inline penn/laundry.py]: Documentation is going to be needed for status errors at some point. I'll trello it.

**Pull request merged**